polygeist/tools/cgeist/Test/Verification/sycl/undeftype.cpp

```diff
@@ -0,0 +1,14 @@
+// RUN: clang++ -fsycl -fsycl-device-only -O0 -w -emit-mlir -o - %s | FileCheck %s
+
+#include <sycl/sycl.hpp>
+
+template <typename T> SYCL_EXTERNAL void keep(T);
+
+// CHECK-LABEL: func.func @_Z13access_marrayim(
+// CHECK:         call @{{.*}}(%{{.*}}, %{{.*}}) : (!llvm.ptr<struct<(array<256 x i32>)>, 4>, memref<?xi32, 4>) -> ()
+// CHECK:         %{{.*}} = call @{{.*}}(%{{.*}}, %arg1) : (!llvm.ptr<struct<(array<256 x i32>)>, 4>, i64) -> memref<?xi32, 4>
+
+SYCL_EXTERNAL void access_marray(int el, size_t i) {
+  sycl::marray<int, 256> marr(el);
+  keep(marr[i]);
+}
```

Comment:
This test is working. I suspect by defining SYCL types we're only hiding cgeist flaws (as #7695).

Reply:
I am not fully understanding, can you please elaborate?
Should we define `sycl::marray` type in SYCL dialect? If we do, what flaws is it hiding?

There are calls to `_ZN4sycl3_V16marrayIiLm256EEC1ERKi` and `_ZN4sycl3_V16marrayIiLm256EEixEm`, which are part of the SYCL namespace. Why shouldn't we generate `sycl.call` for them?

Are you proposing to generate `sycl.call`, `sycl.constructor`, etc only when the method associated class has a MLIR type in the dialect?

Reply:
We do not have support yet for the `sycl::marray` type. The code at line 12 would cause `cgeist` to create a `sycl.constructor` operation, that operation expects valid sycl types, and so it asserts (correctly). 

The obvious solution is to implement support for the `sycl::marray` type. The approach in this PR is to avoid the assertion and continue... Perhaps this would cause some more SYCL test cases to pass, but in the long term IMO we should implement the type. 

At minimum, if we decide to continue with this PR, we should provide an option to enable this behavior (default false). 
We can then use the option on the SYCL test suite to help us root out problems that are not related to the assertions I mentioned. Opinions ?

Reply:
`sycl::marray` was a bad example because we **should** handle that in the future. On the other hand, for implementation types, we might not want to add every single one of them. In the new example, I'm defining a type in the `sycl` namespace (something even a user might do).

I also changed the PR so that we generate `sycl.call`s for member functions of types we do not have in our dialect.

I'd say we do not need a flag for this because it fixes a problem: constructors of types in the `sycl` namespace will break `cgeist`.